spec/models/async_transaction/base_spec.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe AsyncTransaction::Base, type: :model do
+  describe 'Validation' do
+    let(:transaction1) { create(:address_transaction) }
+    let(:transaction2) { build(:address_transaction) }
+    let(:transaction3) { build(:address_transaction, source: nil) }
+    let(:transaction4) { build(:address_transaction) }
+
+    it 'ensures uniqueness across source and transaction_id' do
+      transaction2.id = transaction1.id
+      expect(transaction2.valid?).to be false
```

Comment:
Can we add an expectation in here for the message that we expect the end user to receive?